cli/azd/cmd/util.go

```diff
@@ -74,32 +75,33 @@ func createAndInitEnvironment(
 	azdCtx *azdcontext.AzdContext,
 	console input.Console,
 	azCli azcli.AzCli,
-) (*environment.Environment, context.Context, error) {
+) (*environment.Environment, error) {
 	if envSpec.environmentName != "" && !environment.IsValidEnvironmentName(envSpec.environmentName) {
 		errMsg := invalidEnvironmentNameMsg(envSpec.environmentName)
 		fmt.Fprint(console.Handles().Stdout, errMsg)
-		return nil, nil, fmt.Errorf(errMsg)
+		return nil, fmt.Errorf(errMsg)
 	}
 
 	if err := ensureValidEnvironmentName(ctx, &envSpec.environmentName, console); err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
 	// Ensure the environment does not already exist:
 	env, err := environment.GetEnvironment(azdCtx, envSpec.environmentName)
 	switch {
 	case errors.Is(err, os.ErrNotExist):
 	case err != nil:
-		return nil, nil, fmt.Errorf("checking for existing environment: %w", err)
+		return nil, fmt.Errorf("checking for existing environment: %w", err)
 	case err == nil:
-		return nil, nil, fmt.Errorf("environment '%s' already exists", envSpec.environmentName)
+		return nil, fmt.Errorf("environment '%s' already exists", envSpec.environmentName)
 	}
 
 	if err := ensureEnvironmentInitialized(ctx, *envSpec, env, console, azCli); err != nil {
-		return nil, nil, fmt.Errorf("initializing environment: %w", err)
+		return nil, fmt.Errorf("initializing environment: %w", err)
 	}
 
-	return env, telemetry.ContextWithEnvironment(ctx, env), nil
+	telemetry.SetGlobalAttributes(fields.StringHashed(fields.SubscriptionIdKey, env.GetSubscriptionId()))
```

Comment:
It feels like our environment object should not need to be concerned with how telemetry is collected? After the middleware is in it may be worth refactoring to pull out the subscription ID in the middleware vs here.

Reply:
Code instrumentation and code itself is usually tied together. For example, if I wanted to measure the project config initialization performance, I'd add a code block here to emit a span. If we switch to a usage model based on multiple events, we would also fire a usage event here saying, "project config initialized with subscription ID".

The code itself is not tied to collection. It is simply annotating that there's an attribute worth noting. The actual execution here is that it would set an in-memory dictionary field.

You're also right that we could move this into a middleware for things that are constructed prior to runtime. However, there will still be cases at runtime where we want to say "feature X" is being used. Example would be if we started tracking the service host / language being used. We want to only emit it when the feature is being referenced in the command, i.e. if the user passes `azd deploy --service serviceX`, we want only language/host for `serviceX` to be emitted in telemetry, and not `serviceY`.